### PR TITLE
<Badge/> - expose a more performant API for customization

### PR DESCRIPTION
**Pull request by shlomitc**
This PR is a follow up to this initial POC: https://github.com/wix/wix-ui-tpa/pull/452
1. Expose the `badgeOverrides` public API for performant customization with backward compatibility
2. Allow overriding component variation by eliminating the `-st-state` in favor of a raw class)
3. Improve docs

@sivanharel - bullet 2 is a best practice to apply to all of our components and could be relevant in WSR to make theming easier with shorter selectors (cc @EgoziE )

### CODE REVIEW

**Comment by WixBuildServer** [comment]
### 🤖 Hi, I'm #Janet!

  A deploy preview of your PR was successfully released 🖼.

  You can find it under one of the following URLs:
  * https://bo.wix.com/pages/wix-ui-tpa.pr_456/

File: src/components/Badge/docs/BadgeExtendedExample.st.css [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
         BadgeBgColor '"color(--badgeBgColor)"',
         BadgeTextColor '"color(--badgeTextColor)"',
-        BadgeBorderColor '"color(--badgeBorderColor)"'
     );
-}
\ No newline at end of file
+}
+
+.mixOverrides {
+    -st-mixin: badgeOverrides(
```
Comment (shlomitc) [inline src/components/Badge/docs/BadgeExtendedExample.st.css:15]: @jonathanadler - what do you think of the API? Think that this should become some kind of a convention for all the components we will create

File: src/components/Badge/Badge.tsx [hunk 2]
```diff
@@ -26,8 +26,13 @@ class Badge extends React.Component<BadgeProps> {
     const { priority, children, className, icon } = this.props;
     return (
       <div
+        className={st(
+          classes.root,
+          { withIcon: !!icon },
+          classes[`priority-${priority}`],
+          className,
+        )}
```
Comment (jonathanadler) [inline src/components/Badge/Badge.tsx:34]: So, if I understand correct, what we're actually saying here, is that we want to forbid the use of "variant" states, and allow only booleans.
I wonder though, how can we enforce this?
Some kind of lint rule?

**Review by jonathanadler — COMMENTED** [review]
I think this is awesome
The only thing I'm "concerned" about, is that it will be difficult to enforce.
Or maybe not enforce, but, we'll need to find an easy way for contributors to understand this approach.
I'm thinking maybe we should separate the stylesheet to two: `Badge.st.css` & `overrides.st.css` (also in the generator).
Maybe this will make it clearer?
It will affect extension as well:
```
:import {
  -st-from: "path/to/Badge/overrides.st.css";
  -st-named: tbd;
}
```

File: src/components/Badge/Badge.tsx (hunk 2 above)
Comment (shlomitc) [reply src/components/Badge/Badge.tsx:34]: I just think that we've using stylable wrong all along.
state are there to represents, well, logic states. styles variations are class variations and not states.

so I don't think we need a rule, but rather pass the knowledge and fix ourselves

**Comment by shlomitc** [comment]
> I'm thinking maybe we should separate the stylesheet to two: `Badge.st.css` & `overrides.st.css` (also in the generator).
> 
> ```
> :import {
>   -st-from: "path/to/Badge/overrides.st.css";
>   -st-named: tbd;
> }
> ```

I'm afraid this will not be possible in a way that preserves the existing styling mechanism.
If we separate it to a different file, the css specificity will be different and we will be able to achieve only one of the options:
- only mix the entire file (current behavior, so we did nothing)
- only mix the overrides (the new behavior, but a breaking change)

The only way to go with your approach will be also separating the `priority` themes from the root file. meaning that we will treat each one as a separate variant: `<Badge className={primaryBadge}/>` or `<Badge className={secondaryBadge}/>`.
And then maybe (didn't check yet) the consumer could do `<Badge className={st(primaryBadge, styleParams}/>`. We might be able to do it for new components, but not for the existing ones.

**Pull request merged**